stream/src/main/scala/org/apache/pekko/stream/impl/JavaStreamSource.scala

```diff
@@ -27,28 +30,32 @@ import pekko.stream.stage.{ GraphStage, GraphStageLogic, OutHandler }
   override val shape: SourceShape[T] = SourceShape(out)
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
-    new GraphStageLogic(shape) with OutHandler {
-      private[this] var stream: java.util.stream.BaseStream[T, S] = _
-      private[this] var iter: java.util.Iterator[T] = _
-
-      setHandler(out, this)
+    new GraphStageLogic(shape) with OutHandler with Consumer[T] {
+      private var stream: java.util.stream.BaseStream[T, S] = _
+      private var splitIterator: Spliterator[T] = _
 
-      override def preStart(): Unit = {
+      final override def preStart(): Unit = {
         stream = open()
-        iter = stream.iterator()
+        splitIterator = stream.spliterator()
+        val isEmpty = splitIterator.hasCharacteristics(Spliterator.SIZED) && splitIterator.estimateSize() == 0
+        if (isEmpty) {
```

Comment:
do we need the isEmpty val? couldn't we just have that as the if condition?

Reply:
ok, let me inline it, but it's simpler for read.